### PR TITLE
Use Documenter.jl for documentation (# 70)

### LINKED ISSUE #70: Make proper documentaion
The README is becoming rather encyclopedic. It would be better to have a short README with an intro to the speed improvements over `Array` and the "Quick Start" section with the remaining documentation organized into a documentation webpage.

There are many approaches to take - we could try [Documenter.jl](https://github.com/JuliaDocs/Documenter.jl) combined with a github.io webpage.

### PR DESCRIPTION
**Pull request by jw3126**
Moved most of the `README.md` into `Documenter.jl` documentation, fixes #70 . You can build it locally via `cd docs && julia make.jl`. For hosting someone with appropriate rights in JuliaArrays organization needs to do some [setup](https://juliadocs.github.io/Documenter.jl/stable/man/hosting/).

### CODE REVIEW

**Comment by tkoolen** [comment]
A little gotcha that I ran into: you need to make sure that version tags get built on CI, otherwise you'll only get 'latest' documentation, no 'stable' documentation. Currently, StaticArrays [only has Travis enabled for master](https://github.com/JuliaArrays/StaticArrays.jl/blob/c7d2b2d964a5fcfea5582d069b1f29c089606eb5/.travis.yml#L13).

I fixed this as follows:
https://github.com/tkoolen/RigidBodyDynamics.jl/blob/b08ec1afc9d051e088dac48e221549092157c611/.travis.yml#L10

**Comment by Evizero** [comment]
if CI deployment is a goal then `make.jl` will need some form of `deploydocs`

File: README.md [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 **StaticArrays** provides a framework for implementing statically sized arrays
 in Julia (≥ 0.5), using the abstract type `StaticArray{Size,T,N} <: AbstractArray{T,N}`.
-Subtypes of `StaticArray` will provide fast implementations of common array and
+Subtypes of [StaticArray](@ref) will provide fast implementations of common array and
```
Comment (c42f) [inline README.md]: What is this `@ref` for?

File: docs/src/index.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+Subtypes of [`StaticArray`](@ref) will provide fast implementations of common array and
+linear algebra operations. Note that here "statically sized" means that the
+size can be determined from the *type*, and "static" does **not** necessarily
+imply `immutable`.
+
+The package also provides some concrete static array types: [`SVector`](@ref), [`SMatrix`](@ref)
+and [`SArray`](@ref), which may be used as-is (or else embedded in your own type).
+Mutable versions [`MVector`](@ref), [`MMatrix`](@ref) and [`MArray`](@ref) are also exported, as well
+as [`SizedArray`](@ref) for annotating standard `Array`s with static size information.
+Further, the abstract [`FieldVector`](@ref) can be used to make fast static vectors`s
```
Comment (c42f) [inline docs/src/index.md]: Stray backtick here in ```vectors`s```?

File: README.md (hunk 1 above)
Comment (jw3126) [reply README.md]: It inserts a link to the documentation of the `StaticArray` type.

Comment (c42f) [reply README.md]: Right, but is the README.md actually used during the Documenter build process?  I would have thought the README shouldn't necessarily go in the full docs, and it doesn't appear to be mentioned in docs/make.jl.  (I need to run Documenter for myself, just haven't got around to it yet)

Comment (mschauer) [reply README.md]: Plain documenter ignores README.md, and this link is documenter markup, not markup.

Comment (jw3126) [reply README.md]: Ah I didn't see that this was the README file. Thanks for pointing out!

**Comment by jw3126** [comment]
bump

**Comment by andyferris** [comment]
So, if we merge this will it just work? At the moment I'm getting 404's.

**Comment by tkoolen** [comment]
Actually, I think there's an issue. I just checked the Travis 0.6 Linux log, and there's a segfault while running `make.jl`. Do note that docs will only be deployed when Travis is run on the master branch (i.e. after the PR is merged).

**Comment by tkoolen** [comment]
Running locally, the segfault appears to happen with some probability. I've seen three possible outcomes of running `julia make.jl` from the docs directory:
* everything works
* segfault as on Travis
* an UndefRefError (rarely):

```
Documenter: setting up build directory.
Documenter: expanding markdown templates.
ERROR: LoadError: UndefRefError: access to undefined reference
Stacktrace:
 [1] getindex(::SimpleVector, ::Int64) at ./essentials.jl:252
 [2] next at ./iterators.jl:63 [inlined]
 [3] show_datatype(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::DataType) at ./show.jl:228
 [4] show(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::DataType) at ./show.jl:211
 [5] print(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Type{T} where T) at ./strings/io.jl:29
 [6] print(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Char, ::Type{T} where T<:Tuple, ::Vararg{Type{T} where T<:Tuple,N} where N) at ./strings/io.jl:40
 [7] print_signature(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Type) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Utilities/Utilities.jl:260
 [8] #print_to_string#229(::Void, ::Function, ::Documenter.Utilities.Object, ::Vararg{Documenter.Utilities.Object,N} where N) at ./strings/io.jl:102
 [9] (::Documenter.Expanders.##3#4)(::Tuple{Module,String,Symbol,Documenter.Utilities.Object,Bool,Base.Docs.DocStr}, ::Tuple{Module,String,Symbol,Documenter.Utilities.Object,Bool,Base.Docs.DocStr}) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Expanders.jl:375
 [10] sort!(::Array{Any,1}, ::Int64, ::Int64, ::Base.Sort.InsertionSortAlg, ::Base.Order.Lt{Documenter.Expanders.##3#4}) at ./sort.jl:252
 [11] sort!(::Array{Any,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.Lt{Documenter.Expanders.##3#4}, ::Array{Any,1}) at ./sort.jl:341
 [12] sort!(::Array{Any,1}, ::Base.Sort.MergeSortAlg, ::Base.Order.Lt{Documenter.Expanders.##3#4}) at ./sort.jl:435
 [13] #sort!#7(::Base.Sort.MergeSortAlg, ::Function, ::Function, ::Bool, ::Base.Order.ForwardOrdering, ::Function, ::Array{Any,1}) at ./sort.jl:494
 [14] (::Base.#kw##sort!)(::Array{Any,1}, ::Base.#sort!, ::Array{Any,1}) at ./<missing>:0
 [15] runner(::Type{Documenter.Expanders.AutoDocsBlocks}, ::Base.Markdown.Code, ::Documenter.Documents.Page, ::Documenter.Documents.Document) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Expanders.jl:377
 [16] dispatch(::Type{Documenter.Expanders.ExpanderPipeline}, ::Base.Markdown.Code, ::Documenter.Documents.Page, ::Documenter.Documents.Document) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Selectors.jl:164
 [17] expand(::Documenter.Documents.Document) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Expanders.jl:29
 [18] dispatch(::Type{Documenter.Builder.DocumentPipeline}, ::Documenter.Documents.Document) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Selectors.jl:164
 [19] cd(::Documenter.##2#3{Documenter.Documents.Document}, ::String) at ./file.jl:70
 [20] #makedocs#1(::Bool, ::Array{Any,1}, ::Function) at /home/twan/code/RigidBodyDynamics/v0.6/Documenter/src/Documenter.jl:198
 [21] (::Documenter.#kw##makedocs)(::Array{Any,1}, ::Documenter.#makedocs) at ./<missing>:0
 [22] include_from_node1(::String) at ./loading.jl:569
 [23] include(::String) at ./sysimg.jl:14
 [24] process_options(::Base.JLOptions) at ./client.jl:305
 [25] _start() at ./client.jl:371
while loading /home/twan/code/RigidBodyDynamics/v0.6/StaticArrays/docs/make.jl, in expression starting on line 3
```

**Comment by andyferris** [comment]
Hmm... not sure what to make of that stacktrace!

How should we proceed?

**Comment by jw3126** [comment]
I also get a warning which I do not understand:
```
 > checking for missing docstrings.
 !! 1 docstring potentially missing:

    StaticArrays.FixedSizeArrays
```
However there is a docstring and it can be accessed from julia without problem:
```julia
help?> StaticArrays.FixedSizeArrays
  Compatibility layer for transferring from FixedSizeArrays. This provides alternative definitions of Vec, Mat, Point, FixedVectorNoTuple, @fsa, etc,
  using StaticArrays as a backend.

  The type definitions are not "perfect" matches because the type parameters are different. However, it should cover common method signatures and
  constructors.
```

**Comment by tkoolen** [comment]
> I also get a warning which I do not understand:

That might be because of
https://github.com/JuliaArrays/StaticArrays.jl/pull/298/files#diff-79c035eaebc79219f4a10bb20783c475R312, which filters out docstrings that aren't in the `StaticArrays` module.

It's not saying that `StaticArrays.FixedSizeArrays` doesn't have a docstring, it's saying that the docstring doesn't make it into the final documentation.

**Comment by tkoolen** [comment]
> How should we proceed?

I haven't ever seen this before; definitely hold off on merging this of course. Maybe there's something strange in one of the docstrings that Documenter parses that triggers this?

**Comment by mschauer** [comment]
> > checking for missing docstrings.
 !! 1 docstring potentially missing: StaticArrays.FixedSizeArrays

should read "there is a docstring in the program which is not yet placed into the documentation"

**Comment by Evizero** [comment]
In a different package, where I do use Documenter, I also get a segfault every now and then when running make.

**Comment by tkoolen** [comment]
Running locally, this happened quite frequently (I estimate ~75% percent of the time).

**Comment by jw3126** [comment]
If I delete the following docstring
```
diff --git a/src/util.jl b/src/util.jl
index bb14b1b..b2ae96d 100644
--- a/src/util.jl
+++ b/src/util.jl
@@ -76,17 +76,5 @@ TrivialView(a::AbstractArray{T,N}) where {T,N} = TrivialView{typeof(a),T,N}(a)
 
 # Remove the static dimensions from an array
 
-"""
-    drop_sdims(a)
-
-Return an `AbstractArray` with the same elements as `a`, but with static
-dimensions removed (ie, not a `StaticArray`).
-
-This is useful if you want to override dispatch to call the `Base` version of
-operations such as `kron` instead of the implementation in `StaticArrays`.
-Normally you shouldn't need to do this, but it can be more efficient for
-certain algorithms where the number of elements of the output is a lot larger
-than the input.
-"""
 @inline drop_sdims(a::StaticArray) = TrivialView(a)
 @inline drop_sdims(a) = a
```
I don't get segfaults anymore. Can you confirm @tkoolen? Any idea why this string is problematic?

**Comment by jw3126** [comment]
Now this is strange. If I replace the above docstring by "asdfad" I still get segfaults. Also removing `@inline` does not help.

**Comment by andyferris** [comment]
That stacktrace totally looks like it should never happen. Makes me wonder if it's a weird Julia bug - why would there be an undefined reference in a `DataType.parameters` simple vector??

Would you like to submit an issue to Julia?

**Comment by jw3126** [comment]
As a temporary workaround I commented the offending docstring out. Locally I did not get a single segfault in 20 builds.

**Comment by tkoolen** [comment]
Looks like lots of progress is being made on https://github.com/JuliaLang/julia/issues/22842, so hopefully this will be resolved soon on Julia nightly. I think the workaround is fine for now.

**Comment by c42f** [comment]
Hey guys, I've added a deploy key to StaticArrays on github and Travis, which will hopefully be enough to make this work once it's merged.

**Comment by c42f** [comment]
I also did a little extra testing locally which seemed to work, so I guess we merge this and see what happens.

**Comment by c42f** [comment]
Well, the documentation didn't seem to deploy from Travis after merging this (ie, nothing got pushed to to the gh-pages branch). Does anyone know why this might be? I've set `DOCUMENTER_DEBUG=true` and restarted the build to see some more information.

**Comment by tkoolen** [comment]
Looks like doctests failed (expand line 879 of https://travis-ci.org/JuliaArrays/StaticArrays.jl/jobs/278842786); not sure if that causes docs to not be deployed, but should be fixed anyway.

I think the main reason is `deploying if equal to "nightly"` in your debug build.

**Comment by tkoolen** [comment]
In addition to `julia = "0.6"`, also add the following kwargs to `deploydocs`:
```julia
    target = "build",
    deps   = nothing,
    make   = nothing
```
as per https://juliadocs.github.io/Documenter.jl/stable/man/guide/#Output-formats-1 since, 'Native HTML Output' box.

**Comment by tkoolen** [comment]
Another mistake: the `repo` arg of `deploydocs` should be 
```julia
repo = "github.com/JuliaArrays/StaticArrays.jl.git"
```
(the `.git` is missing).

**Comment by jw3126** [comment]
I fixed these and pushed to the docs to the same branch as this PR. Do I need to open a new PR or can this PR be reopened somehow?

**Comment by tkoolen** [comment]
New PR.

**Comment by c42f** [comment]
Hmm.  According to the Documenter docs it's a straightforward rules problem in the call to `deploydocs`, hopefully will be addressed by #305.

**Comment by c42f** [comment]
Oops I didn't see the several comments above before making #305. I don't think it addresses all the problems yet, so if you've got something complete @jw3126 feel free to open a separate PR and I'll merge it instead.

**Comment by jw3126** [comment]
@c42f Just go ahead, its better when you do it, since I cannot test the deployment. I only added @tkoolen suggestions, nothing else.

**Comment by andyferris** [comment]
Thank you very much to everyone involved in getting this done - it's really great to see this live!

**Pull request merged**